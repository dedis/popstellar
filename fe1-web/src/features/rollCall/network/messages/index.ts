export * from './CloseRollCall';
export * from './CreateRollCall';
export * from './OpenRollCall';
export * from './ReopenRollCall';
