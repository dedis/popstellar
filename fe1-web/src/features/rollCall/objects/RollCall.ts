import {
  Hash,
  HashState,
  PopToken,
  PublicKey,
  PublicKeyState,
  Timestamp,
  TimestampState,
} from 'core/objects';

/**
 * Object to represent a roll call.
 */

export enum RollCallStatus {
  CREATED,
  OPENED,
  CLOSED,
  REOPENED,
}

export interface RollCallState {
  id: HashState;
  idAlias?: HashState;
  name: string;
  location: string;
  description?: string;
  creation: TimestampState;
  proposedStart: TimestampState;
  proposedEnd: TimestampState;
  openedAt?: TimestampState;
  closedAt?: TimestampState;
  status: number;
  attendees?: PublicKeyState[];
}

export class RollCall {
  public static EVENT_TYPE = 'ROLL_CALL';

  public readonly id: Hash;

  public readonly idAlias?: Hash;

  public readonly name: string;

  public readonly location: string;

  public readonly description?: string;

  public readonly creation: Timestamp;

  public readonly proposedStart: Timestamp;

  public readonly proposedEnd: Timestamp;

  public readonly openedAt?: Timestamp;

  public readonly closedAt?: Timestamp;

  public readonly status: RollCallStatus;

  public readonly attendees?: PublicKey[];

  public get start() {
    return this.openedAt || this.proposedStart;
  }

  public get end() {
    return this.closedAt;
  }

  /* Not yet implemented:
   * This object should probably also keep a list of the time periods
   * during which the roll call was open (or re-opened).
   */

  constructor(obj: Partial<RollCall>) {
    if (obj === undefined || obj === null) {
      throw new Error(
        'Error encountered while creating a RollCall object: undefined/null parameters',
      );
    }

    if (obj.id === undefined) {
      throw new Error("Undefined 'id' when creating 'RollCall'");
    }
    if (obj.name === undefined) {
      throw new Error("Undefined 'name' when creating 'RollCall'");
    }
    if (obj.location === undefined) {
      throw new Error("Undefined 'location' when creating 'RollCall'");
    }
    if (obj.creation === undefined) {
      throw new Error("Undefined 'creation' when creating 'RollCall'");
    }
    if (obj.proposedStart === undefined) {
      throw new Error("Undefined 'proposed_start' when creating 'RollCall'");
    }
    if (obj.proposedEnd === undefined) {
      throw new Error("Undefined 'proposed_end' when creating 'RollCall'");
    }
    if (obj.status === undefined) {
      throw new Error("Undefined 'status' when creating 'RollCall'");
    }

    this.id = obj.id;
    this.idAlias = obj.idAlias;
    this.name = obj.name;
    this.location = obj.location;
    this.description = obj.description;
    this.creation = obj.creation;
    this.proposedStart = obj.proposedStart;
    this.proposedEnd = obj.proposedEnd;
    this.status = obj.status;
    this.attendees = obj.attendees;

    this.openedAt = obj.openedAt;
    this.closedAt = obj.closedAt;
  }

  /**
   * Checks if a pop token is contained in the current roll call.
   *
   * @param token - The pop token to search for
   * @returns A boolean telling if the token is there or not
   */
  public containsToken(token: PopToken | undefined): boolean {
    if (this.attendees === undefined || token === undefined) {
      return false;
    }

    return this.attendees.some((attendee: PublicKey) => attendee.equals(token.publicKey));
  }

  /**
   * Creates a RollCallState object from the current RollCall object.
   */
  public toState(): RollCallState {
    return {
      id: this.id.toState(),
      idAlias: this.idAlias?.toState(),
      name: this.name,
      location: this.location,
      description: this.description,
      creation: this.creation.toState(),
      proposedStart: this.proposedStart.toState(),
      proposedEnd: this.proposedEnd.toState(),
      openedAt: this.openedAt?.toState(),
      closedAt: this.closedAt?.toState(),
      status: this.status,
      attendees: this.attendees?.map((attendee) => attendee.toState()),
    };
  }

  /**
   * Creates a RollCall object from a RollCallState object.
   */
  public static fromState(rollCallState: RollCallState): RollCall {
    return new RollCall({
      id: Hash.fromState(rollCallState.id),
      idAlias: rollCallState.idAlias ? Hash.fromState(rollCallState.idAlias) : undefined,
      name: rollCallState.name,
      location: rollCallState.location,
      description: rollCallState.description,
      creation: Timestamp.fromState(rollCallState.creation),
      proposedStart: Timestamp.fromState(rollCallState.proposedStart),
      proposedEnd: Timestamp.fromState(rollCallState.proposedEnd),
      openedAt:
        rollCallState.openedAt !== undefined
          ? Timestamp.fromState(rollCallState.openedAt)
          : undefined,
      closedAt:
        rollCallState.closedAt !== undefined
          ? Timestamp.fromState(rollCallState.closedAt)
          : undefined,
      status: rollCallState.status,
      attendees: rollCallState.attendees?.map((a) => PublicKey.fromState(a)),
    });
  }
}
