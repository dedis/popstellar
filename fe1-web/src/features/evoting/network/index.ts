import { ActionType, ObjectType } from 'core/network/jsonrpc/messages';

import { EvotingConfiguration } from '../interface';
import {
  handleCastVoteMessage,
  handleElectionEndMessage,
  handleElectionOpenMessage,
  handleElectionResultMessage,
  handleElectionSetupMessage,
} from './ElectionHandler';
import { CastVote, ElectionResult, EndElection, SetupElection } from './messages';
import { OpenElection } from './messages/OpenElection';

/**
 * Configures the network callbacks in a MessageRegistry.
 * @param config - An evoting config object
 */
export const configureNetwork = (config: EvotingConfiguration) => {
  config.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.SETUP,
    handleElectionSetupMessage(config.addEvent),
    SetupElection.fromJson,
  );
  config.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.OPEN,
    handleElectionOpenMessage(config.getEventById, config.updateEvent),
    OpenElection.fromJson,
  );
  config.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.CAST_VOTE,
    handleCastVoteMessage(config.getCurrentLao, config.getEventById, config.updateEvent),
    CastVote.fromJson,
  );
  config.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.END,
    handleElectionEndMessage(config.getEventById, config.updateEvent),
    EndElection.fromJson,
  );
  config.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.RESULT,
    handleElectionResultMessage(config.getEventById, config.updateEvent),
    ElectionResult.fromJson,
  );
};
