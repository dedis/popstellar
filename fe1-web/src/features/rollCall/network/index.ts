import { ActionType, MessageRegistry, ObjectType } from 'core/network/jsonrpc/messages';

import { CloseRollCall, CreateRollCall, OpenRollCall, ReopenRollCall } from './messages';
import {
  handleRollCallCloseMessage,
  handleRollCallCreateMessage,
  handleRollCallOpenMessage,
  handleRollCallReopenMessage,
} from './RollCallHandler';

export * from './RollCallMessageApi';

/**
 * Configures the network callbacks in a MessageRegistry.
 *
 * @param registry - The MessageRegistry where we want to add the mappings
 */
export function configureNetwork(registry: MessageRegistry) {
  registry.add(
    ObjectType.ROLL_CALL,
    ActionType.CREATE,
    handleRollCallCreateMessage,
    CreateRollCall.fromJson,
  );
  registry.add(
    ObjectType.ROLL_CALL,
    ActionType.OPEN,
    handleRollCallOpenMessage,
    OpenRollCall.fromJson,
  );
  registry.add(
    ObjectType.ROLL_CALL,
    ActionType.CLOSE,
    handleRollCallCloseMessage,
    CloseRollCall.fromJson,
  );
  registry.add(
    ObjectType.ROLL_CALL,
    ActionType.REOPEN,
    handleRollCallReopenMessage,
    ReopenRollCall.fromJson,
  );
}
