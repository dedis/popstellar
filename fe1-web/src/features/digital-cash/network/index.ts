import { ActionType, MessageRegistry, ObjectType } from 'core/network/jsonrpc/messages';
import { Hash, PublicKey } from 'core/objects';
import { dispatch } from 'core/redux';

import { Transaction } from '../objects/transaction';
import { addTransaction } from '../reducer';
import { handleTransactionPost } from './DigitalCashHandler';
import { PostTransaction } from './messages';

export * from './DigitalCashMessageApi';

/**
 * Configures the network callbacks in a MessageRegistry.
 *
 * @param registry - The MessageRegistry where we want to add the mappings
 * @param getLaoOrganizer - A function to get the organizer from a lao id
 */
export function configureNetwork(
  registry: MessageRegistry,
  getLaoOrganizer: (laoId: Hash) => PublicKey | undefined,
) {
  const addTransactionToState = (laoId: Hash, transaction: Transaction) => {
    dispatch(addTransaction(laoId, transaction));
  };

  registry.add(
    ObjectType.COIN,
    ActionType.POST_TRANSACTION,
    handleTransactionPost(addTransactionToState, getLaoOrganizer),
    PostTransaction.fromJSON,
  );
}
