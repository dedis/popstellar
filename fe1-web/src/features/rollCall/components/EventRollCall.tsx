import { useNavigation } from '@react-navigation/native';
import PropTypes from 'prop-types';
import React, { useEffect, useState, FunctionComponent } from 'react';
import { Text } from 'react-native';
import { useToast } from 'react-native-toast-notifications';
import { useSelector } from 'react-redux';

import { QRCode, WideButtonView } from 'core/components';
import { selectCurrentLao } from 'features/lao/reducer';
import * as Wallet from 'features/wallet/objects';
import { WalletStore } from 'features/wallet/store';
import { FOUR_SECONDS } from 'resources/const';
import STRINGS from 'resources/strings';

import { requestOpenRollCall, requestReopenRollCall } from '../network';
import { RollCall, RollCallStatus } from '../objects';

/**
 * Component used to display a RollCall event in the LAO event list
 */
const EventRollCall = (props: IPropTypes) => {
  const { event: rollCall } = props;
  const { isOrganizer } = props;
  const lao = useSelector(selectCurrentLao);
  // FIXME: use a more specific navigation
  const navigation = useNavigation<any>();
  const toast = useToast();

  if (!lao) {
    throw new Error('no LAO is currently active');
  }
  const [popToken, setPopToken] = useState('');
  const [hasWalletBeenInitialized, setHasWalletBeenInitialized] = useState(WalletStore.hasSeed());

  // re-check if wallet has been initialized after focus events
  useEffect(() => {
    // Return the function to unsubscribe from the event so it gets removed on unmount
    return navigation.addListener('focus', () => {
      setHasWalletBeenInitialized(WalletStore.hasSeed());
    });
  }, [navigation]);

  // Once the roll call is opened the first time, idAlias is defined, and needed for closing/reopening the roll call
  const eventHasBeenOpened = rollCall.idAlias !== undefined;

  useEffect(() => {
    if (!hasWalletBeenInitialized || !lao?.id || !rollCall?.id) {
      return;
    }

    // Here we get the pop-token to display in the QR code
    Wallet.generateToken(lao.id, rollCall.id)
      .then((token) => setPopToken(token.publicKey.valueOf()))
      .catch((err) => console.error(`Could not generate token: ${err}`));
  }, [hasWalletBeenInitialized, lao, rollCall]);

  if (!rollCall) {
    console.debug('Error in Roll Call display: Roll Call doesnt exist in store');
    return null;
  }

  const makeToastErr = (error: string) => {
    toast.show(error, {
      type: 'danger',
      placement: 'bottom',
      duration: FOUR_SECONDS,
    });
  };

  const onOpenRollCall = () => {
    requestOpenRollCall(rollCall.id).catch((e) => {
      makeToastErr('Unable to send roll call open request');
      console.debug('Unable to send Roll call open request', e);
    });
  };

  const onReopenRollCall = () => {
    if (eventHasBeenOpened) {
      requestReopenRollCall(rollCall.idAlias).catch((e) => {
        makeToastErr('Unable to send Roll call re-open request');
        console.debug('Unable to send Roll call re-open request', e);
      });
    } else {
      makeToastErr('Unable to send roll call re-open request, the event does not have an idAlias');
      console.debug('Unable to send roll call re-open request, the event does not have an idAlias');
    }
  };

  const onScanAttendees = () => {
    if (eventHasBeenOpened) {
      navigation.navigate(STRINGS.roll_call_open, {
        rollCallID: rollCall.id.toString(),
      });
    } else {
      makeToastErr('Unable to scan attendees, the event does not have an idAlias');
      console.debug('Unable to scan attendees, the event does not have an idAlias');
    }
  };

  const getRollCallDisplay = (status: RollCallStatus) => {
    switch (status) {
      case RollCallStatus.CREATED:
        return (
          <>
            <Text>Not Open yet</Text>
            <Text>Be sure to have set up your Wallet</Text>
            {isOrganizer && (
              <WideButtonView title={STRINGS.roll_call_open} onPress={() => onOpenRollCall()} />
            )}
          </>
        );
      case RollCallStatus.REOPENED:
      case RollCallStatus.OPENED:
        return (
          <>
            {!isOrganizer && (
              <>
                <Text>Let the organizer scan your Pop token</Text>
                <QRCode visibility value={popToken} />
              </>
            )}
            {isOrganizer && (
              <WideButtonView
                title={STRINGS.roll_call_scan_attendees}
                onPress={() => onScanAttendees()}
              />
            )}
          </>
        );
      case RollCallStatus.CLOSED:
        return (
          <>
            <Text>Closed</Text>
            <Text>Attendees are:</Text>
            {rollCall.attendees?.map((attendee) => (
              <Text key={attendee.valueOf()}>{attendee}</Text>
            ))}
            {isOrganizer && (
              <WideButtonView title={STRINGS.roll_call_reopen} onPress={() => onReopenRollCall()} />
            )}
          </>
        );
      default:
        console.warn('Roll Call Status was undefined in EventRollCall');
        return null;
    }
  };

  return (
    <>
      <Text>Roll Call</Text>
      {getRollCallDisplay(rollCall.status)}
    </>
  );
};

const propTypes = {
  event: PropTypes.instanceOf(RollCall).isRequired,
  isOrganizer: PropTypes.bool,
};
EventRollCall.propTypes = propTypes;

EventRollCall.defaultProps = {
  isOrganizer: false,
};

type IPropTypes = PropTypes.InferProps<typeof propTypes>;

export default EventRollCall;

export const RollCallEventTypeComponent = {
  isOfType: (event: unknown) => event instanceof RollCall,
  Component: EventRollCall as FunctionComponent<{
    event: unknown;
    isOrganizer: boolean | null | undefined;
  }>,
};
