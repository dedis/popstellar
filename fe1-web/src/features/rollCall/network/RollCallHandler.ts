import { subscribeToChannel } from 'core/network';
import { ActionType, ObjectType, ProcessableMessage } from 'core/network/jsonrpc/messages';
import { getCoinChannel, getReactionChannel, getUserSocialChannel, Hash } from 'core/objects';
import { AsyncDispatch, dispatch } from 'core/redux';

import { RollCallConfiguration } from '../interface';
import { RollCall, RollCallStatus } from '../objects';
import { CloseRollCall, CreateRollCall, OpenRollCall, ReopenRollCall } from './messages';

/**
 * Handles a RollCallCreate message by creating a roll call in the current Lao.
 * @param addRollCall - A function to add a new roll call
 */
export const handleRollCallCreateMessage =
  (addRollCall: (laoId: Hash, rollCall: RollCall) => void) =>
  (msg: ProcessableMessage): boolean => {
    const makeErr = (err: string) => `roll_call#create was not processed: ${err}`;

    if (
      msg.messageData.object !== ObjectType.ROLL_CALL ||
      msg.messageData.action !== ActionType.CREATE
    ) {
      console.warn(
        makeErr(
          `Invalid object or action parameter: ${msg.messageData.object}#${msg.messageData.action}`,
        ),
      );
      return false;
    }

    if (!msg.laoId) {
      console.warn(makeErr(`Was not sent on a lao subchannel but rather on '${msg.channel}'`));
      return false;
    }

    const rcMsgData = msg.messageData as CreateRollCall;

    const rollCall = new RollCall({
      id: rcMsgData.id,
      name: rcMsgData.name,
      location: rcMsgData.location,
      description: rcMsgData.description,
      creation: rcMsgData.creation,
      proposedStart: rcMsgData.proposed_start,
      proposedEnd: rcMsgData.proposed_end,
      status: RollCallStatus.CREATED,
    });

    addRollCall(msg.laoId, rollCall);
    return true;
  };

/**
 * Handle a RollCallOpen message by opening the corresponding roll call.
 * @param getRollCallById - A function to get a roll call by its id
 * @param updateRollCall - An function to update a roll call
 */
export const handleRollCallOpenMessage =
  (
    getRollCallById: (rollCallId: Hash) => RollCall | undefined,
    updateRollCall: (rollCall: RollCall) => void,
  ) =>
  (msg: ProcessableMessage): boolean => {
    const makeErr = (err: string) => `roll_call#open was not processed: ${err}`;

    if (
      msg.messageData.object !== ObjectType.ROLL_CALL ||
      msg.messageData.action !== ActionType.OPEN
    ) {
      console.warn(
        makeErr(
          `Invalid object or action parameter: ${msg.messageData.object}#${msg.messageData.action}`,
        ),
      );
      return false;
    }

    if (!msg.laoId) {
      console.warn(makeErr(`Was not sent on a lao subchannel but rather on '${msg.channel}'`));
      return false;
    }

    const rcMsgData = msg.messageData as OpenRollCall;
    const oldRC = getRollCallById(rcMsgData.opens);
    if (!oldRC) {
      console.warn(makeErr("no known roll call matching the 'opens' field"));
      return false;
    }

    const rollCall = new RollCall({
      ...oldRC,
      idAlias: rcMsgData.update_id,
      openedAt: rcMsgData.opened_at,
      status: RollCallStatus.OPENED,
    });

    updateRollCall(rollCall);
    return true;
  };

/**
 * Handles a RollCallClose message by closing the corresponding roll call.
 * @param getRollCallById - A function to get a roll call by its id
 * @param updateRollCall - An function to update a roll call
 * @param generateToken - A function to generate a pop token
 * @param setLaoLastRollCall - An action creator to set the last (tokenized) roll call for a given lao
 */
export const handleRollCallCloseMessage =
  (
    getRollCallById: (rollCallId: Hash) => RollCall | undefined,
    updateRollCall: (rollCall: RollCall) => void,
    generateToken: RollCallConfiguration['generateToken'],
    setLaoLastRollCall: RollCallConfiguration['setLaoLastRollCall'],
  ) =>
  (msg: ProcessableMessage): boolean => {
    const makeErr = (err: string) => `roll_call#close was not processed: ${err}`;

    if (
      msg.messageData.object !== ObjectType.ROLL_CALL ||
      msg.messageData.action !== ActionType.CLOSE
    ) {
      console.warn(
        makeErr(
          `Invalid object or action parameter: ${msg.messageData.object}#${msg.messageData.action}`,
        ),
      );
      return false;
    }

    if (!msg.laoId) {
      console.warn(makeErr(`Was not sent on a lao subchannel but rather on '${msg.channel}'`));
      return false;
    }

    const { laoId } = msg;

    const rcMsgData = msg.messageData as CloseRollCall;
    const oldRC = getRollCallById(rcMsgData.closes);
    if (!oldRC) {
      console.warn(makeErr("no known roll call matching the 'closes' field"));
      return false;
    }

    const rollCall = new RollCall({
      ...oldRC,
      idAlias: rcMsgData.update_id,
      closedAt: rcMsgData.closed_at,
      status: RollCallStatus.CLOSED,
      attendees: rcMsgData.attendees,
    });

    updateRollCall(rollCall);

    // ... and update the Lao state to point to the latest roll call, if we have a token in it.
    dispatch(async (aDispatch: AsyncDispatch) => {
      try {
        const token = await generateToken(laoId, rollCall.id);
        const hasToken = rollCall.containsToken(token);
        aDispatch(setLaoLastRollCall(laoId, rollCall.id, hasToken));

        const promises: Promise<unknown>[] = [];

        if (rollCall.attendees) {
          // subscribe to all roll call attendee's social media channels
          const subscriptionPromises = rollCall.attendees.map((attendee) =>
            subscribeToChannel(laoId, dispatch, getUserSocialChannel(laoId, attendee)).catch(
              (err) => {
                console.error(
                  `Could not subscribe to social channel of attendee with publuc key '${attendee}', error:`,
                  err,
                );
              },
            ),
          );

          // push all promises into the array
          promises.push(...subscriptionPromises);
        }

        // everyone is automatically subscribed to the reaction channel after the roll call
        promises.push(
          subscribeToChannel(laoId, dispatch, getReactionChannel(laoId)).catch((err) => {
            console.error('Could not subscribe to reaction channel, error:', err);
          }),
        );

        // we also subscribe to the coin channel of this roll call
        promises.push(
          subscribeToChannel(laoId, dispatch, getCoinChannel(laoId)).catch((err) => {
            console.error('Could not subscribe to coin channel, error: ', err);
          }),
        );

        // wait until all promises return
        await Promise.all(promises);
      } catch (err) {
        console.debug(err);
      }
    });

    return true;
  };

/**
 * Handles a RollCallReopen message by reopening the corresponding roll call.
 *
 * @param getRollCallById - A function to get a roll call by its id
 * @param updateRollCall - An function to update a roll call
 */
export const handleRollCallReopenMessage =
  (
    getRollCallById: (rollCallId: Hash) => RollCall | undefined,
    updateRollCall: (rollCall: RollCall) => void,
  ) =>
  (msg: ProcessableMessage) => {
    const makeErr = (err: string) => `roll_call#reopen was not processed: ${err}`;

    if (
      msg.messageData.object !== ObjectType.ROLL_CALL ||
      msg.messageData.action !== ActionType.REOPEN
    ) {
      console.warn(
        makeErr(
          `Invalid object or action parameter: ${msg.messageData.object}#${msg.messageData.action}`,
        ),
      );
      return false;
    }

    if (!msg.laoId) {
      console.warn(makeErr(`Was not sent on a lao subchannel but rather on '${msg.channel}'`));
      return false;
    }

    const rcMsgData = msg.messageData as ReopenRollCall;
    const oldRC = getRollCallById(rcMsgData.opens);
    if (!oldRC) {
      console.warn(makeErr("no known roll call matching the 'opens' field"));
      return false;
    }
    if (oldRC.status !== RollCallStatus.CLOSED) {
      console.error(makeErr('The roll call status is not coherent'));
      return false;
    }

    const rollCall = new RollCall({
      ...oldRC,
      idAlias: rcMsgData.update_id,
      openedAt: rcMsgData.opened_at,
      status: RollCallStatus.REOPENED,
      // unset end as the roll call is open once again
      closedAt: undefined,
    });

    updateRollCall(rollCall);
    return true;
  };
