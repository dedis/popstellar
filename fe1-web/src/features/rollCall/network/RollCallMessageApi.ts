import { publish } from 'core/network';
import { channelFromIds, EventTags, Hash, PublicKey, Timestamp } from 'core/objects';

import { CloseRollCall, CreateRollCall, OpenRollCall, ReopenRollCall } from './messages';

/**
 * Contains all functions to send roll call related messages.
 */

/**
 * Sends a server query asking for the creation of a roll call.
 *
 * @param laoId - The id of the lao to create the roll call in
 * @param name - The name of the roll call
 * @param location - The location of the roll call
 * @param proposedStart - The time at which the attendee scanning will start (pending confirmation)
 * @param proposedEnd - The time at which the attendee scanning will stop (pending confirmation)
 * @param description - Further information regarding the event (optional)
 * @return A Promise resolving when the operation is completed
 */
export function requestCreateRollCall(
  laoId: Hash,
  name: string,
  location: string,
  proposedStart: Timestamp,
  proposedEnd: Timestamp,
  description?: string,
): Promise<void> {
  const time: Timestamp = Timestamp.EpochNow();

  const message = new CreateRollCall(
    {
      id: Hash.fromArray(EventTags.ROLL_CALL, laoId, time, name),
      name: name,
      creation: time,
      location: location,
      proposed_start: Timestamp.max(time, proposedStart),
      proposed_end: proposedEnd,
      description: description,
    },
    laoId,
  );

  const laoCh = channelFromIds(laoId);
  return publish(laoCh, message);
}

/**
 * Sends a server query asking for the opening of a roll call.
 *
 * @param laoId - The id of the lao to open the roll call in
 * @param rollCallId - The ID of the roll call to open
 * @param start - The time at which the operation happens, defaults to now (optional)
 * @return A Promise resolving when the operation is completed
 */
export function requestOpenRollCall(
  laoId: Hash,
  rollCallId: Hash,
  start?: Timestamp,
): Promise<void> {
  const time = start === undefined ? Timestamp.EpochNow() : start;

  const message = new OpenRollCall(
    {
      update_id: OpenRollCall.computeOpenRollCallId(laoId, rollCallId, time),
      opens: rollCallId,
      opened_at: time,
    },
    laoId,
  );

  return publish(channelFromIds(laoId), message);
}

/**
 * Sends a server query asking for the re-opening of a roll call.
 *
 * @param laoId - The id of the lao to reopen the roll call in
 * @param rollCallId - The ID of the roll call to re-open
 * @param start - The time at which the operation happens, defaults to now (optional)
 * @return A Promise resolving when the operation is completed
 */
export function requestReopenRollCall(
  laoId: Hash,
  rollCallId: Hash,
  start?: Timestamp,
): Promise<void> {
  const time = start === undefined ? Timestamp.EpochNow() : start;

  const message = new ReopenRollCall(
    {
      update_id: ReopenRollCall.computeOpenRollCallId(laoId, rollCallId, time),
      opens: rollCallId,
      opened_at: time,
    },
    laoId,
  );

  return publish(channelFromIds(laoId), message);
}

/**
 * Sends a server query asking for the closing of a roll call.
 *
 * @param laoId - The id of the lao to close the roll call in
 * @param rollCallId - The ID of the roll call to close
 * @param attendees - The list of the attendees' public keys
 * @param close - The time at which the operation happens, defaults to now (optional)
 * @return A Promise resolving when the operation is completed
 */
export function requestCloseRollCall(
  laoId: Hash,
  rollCallId: Hash,
  attendees: PublicKey[],
  close?: Timestamp,
): Promise<void> {
  const time = close === undefined ? Timestamp.EpochNow() : close;

  const message = new CloseRollCall(
    {
      update_id: CloseRollCall.computeCloseRollCallId(laoId, rollCallId, time),
      closes: rollCallId,
      closed_at: time,
      attendees: attendees,
    },
    laoId,
  );

  return publish(channelFromIds(laoId), message);
}
