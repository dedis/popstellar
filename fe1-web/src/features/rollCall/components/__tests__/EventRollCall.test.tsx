import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import * as reactRedux from 'react-redux';

import { mockNavigate } from '__mocks__/useNavigationMock';
import { mockLao } from '__tests__/utils/TestUtils';
import { Hash, Timestamp } from 'core/objects';
import STRINGS from 'resources/strings';

import { requestOpenRollCall, requestReopenRollCall } from '../../network';
import { EventTypeRollCall, RollCall, RollCallStatus } from '../../objects';
import EventRollCall from '../EventRollCall';

const ID = new Hash('rollCallId');
const NAME = 'myRollCall';
const LOCATION = 'location';
const TIMESTAMP_START = new Timestamp(1620255600);
const TIMESTAMP_END = new Timestamp(1620357600);
const ATTENDEES = ['attendee1', 'attendee2'];

const createStateWithStatus: any = (mockStatus: RollCallStatus) => {
  return {
    id: ID.valueOf(),
    eventType: EventTypeRollCall,
    start: TIMESTAMP_START.valueOf(),
    name: NAME,
    location: LOCATION,
    creation: TIMESTAMP_START.valueOf(),
    proposedStart: TIMESTAMP_START.valueOf(),
    proposedEnd: TIMESTAMP_END.valueOf(),
    status: mockStatus,
    attendees: ATTENDEES,
    idAlias: mockStatus === RollCallStatus.CREATED ? undefined : ID.valueOf(),
  };
};

const mockRollCallClosed = RollCall.fromState(createStateWithStatus(RollCallStatus.CLOSED));
const mockRollCallCreated = RollCall.fromState(createStateWithStatus(RollCallStatus.CREATED));
const mockRollCallOpened = RollCall.fromState(createStateWithStatus(RollCallStatus.OPENED));

jest.mock('@react-navigation/native');
jest.mock('features/rollCall/network', () => ({
  requestOpenRollCall: jest.fn(() => Promise.resolve()),
  requestReopenRollCall: jest.fn(() => Promise.resolve()),
}));

let mockSelector: any;

beforeEach(() => {
  jest.clearAllMocks();
  mockSelector = jest.spyOn(reactRedux, 'useSelector').mockReturnValueOnce(mockLao);
});

describe('EventRollCall', () => {
  it('should correctly render', () => {
    mockSelector.mockReturnValueOnce(mockRollCallCreated);
    const obj = render(<EventRollCall event={mockRollCallCreated} isOrganizer={false} />);
    expect(obj.toJSON()).toMatchSnapshot();
  });

  it('should call requestOpenRollCall when the open button is clicked', () => {
    const usedMockRollCall = mockRollCallCreated;
    mockSelector.mockReturnValueOnce(usedMockRollCall);
    const obj = render(<EventRollCall event={usedMockRollCall} isOrganizer />);
    const openRollCallButton = obj.getByText(STRINGS.roll_call_open);
    fireEvent.press(openRollCallButton);
    expect(requestOpenRollCall).toHaveBeenCalledTimes(1);
  });

  it('should call requestReopenRollCall when the reopen button is clicked', () => {
    const usedMockRollCall = mockRollCallClosed;
    mockSelector.mockReturnValueOnce(usedMockRollCall);
    const obj = render(<EventRollCall event={usedMockRollCall} isOrganizer />);
    const reopenRollCallButton = obj.getByText(STRINGS.roll_call_reopen);
    fireEvent.press(reopenRollCallButton);
    expect(requestReopenRollCall).toHaveBeenCalledTimes(1);
  });

  it('should navigate to RollCallOpened when scan attendees button is clicked', () => {
    const usedMockRollCall = mockRollCallOpened;
    mockSelector.mockReturnValueOnce(usedMockRollCall);
    const obj = render(<EventRollCall event={usedMockRollCall} isOrganizer />);
    const scanAttendeesButton = obj.getByText(STRINGS.roll_call_scan_attendees);
    fireEvent.press(scanAttendeesButton);
    expect(mockNavigate).toHaveBeenCalledTimes(1);
  });
});
