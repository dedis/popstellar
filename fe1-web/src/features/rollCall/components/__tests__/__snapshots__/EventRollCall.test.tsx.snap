// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventRollCall should correctly render 1`] = `
Array [
  <Text>
    Roll Call
  </Text>,
  <Text>
    Not Open yet
  </Text>,
  <Text>
    Be sure to have set up your Wallet
  </Text>,
]
`;
